server/src/main/java/io/crate/metadata/TransportDropSchema.java

```diff
@@ -117,15 +143,67 @@ public ClusterState execute(ClusterState currentState) throws Exception {
             Metadata currentMetadata = currentState.metadata();
             ImmutableOpenMap<String, SchemaMetadata> schemas = currentMetadata.schemas();
             Builder newMetadata = Metadata.builder(currentMetadata);
+            List<RelationName> droppedRelations = new ArrayList<>();
+            List<Index> droppedIndices = new ArrayList<>();
+
             for (String schema : request.names()) {
-                if (!schemas.containsKey(schema) && request.ifExists()) {
-                    continue;
+                SchemaMetadata schemaMetadata = schemas.get(schema);
+                if (schemaMetadata == null) {
+                    if (request.ifExists()) {
+                        continue;
+                    }
+                    throw new SchemaUnknownException(schema);
+                }
+                if (schemaMetadata.relations().isEmpty()) {
+                    newMetadata.dropSchema(schema);
+                } else if (request.cascadeMode() == CascadeMode.RESTRICT) {
+                    throw new DependentObjectsExists("schema", schema);
+                } else {
+                    newMetadata.dropSchema(schema);
+                    for (var cursor : schemaMetadata.relations()) {
+                        RelationMetadata relation = cursor.value;
+                        newMetadata.dropRelation(relation.name());
+                        droppedRelations.add(relation.name());
+                        for (String indexUUID : relation.indexUUIDs()) {
+                            IndexMetadata index = currentMetadata.index(indexUUID);
+                            droppedIndices.add(index.getIndex());
+                        }
+                    }
                 }
-                newMetadata.dropSchema(schema);
             }
-            return ClusterState.builder(currentState)
+
+            ViewsMetadata views = currentMetadata.custom(ViewsMetadata.TYPE);
+            List<RelationName> droppedViews;
+            if (views == null) {
+                droppedViews = List.of();
+            } else {
+                droppedViews = new ArrayList<>();
+                for (String fqViewName : views.names()) {
+                    IndexParts indexParts = IndexName.decode(fqViewName);
+                    RelationName relationName = indexParts.toRelationName();
+                    if (request.names().contains(relationName.schema())) {
+                        if (request.cascadeMode() == CascadeMode.RESTRICT) {
+                            throw new DependentObjectsExists("schema", relationName.schema());
+                        }
+                        droppedViews.add(relationName);
+                    }
+                }
+                if (!droppedViews.isEmpty()) {
+                    RemoveResult remove = views.remove(droppedViews);
+                    newMetadata.putCustom(ViewsMetadata.TYPE, remove.updatedViews());
+                }
```

Comment:
Handling the custom meta types somehow gets messy - would have to do the same for UDF and foreign tables.

I wonder if instead we should document as limitations that explicit create/drop schema currently only considers tables and remove that limitations once we migrated the metadata.